File: apps/ios/Suzent/PairingViews.swift
import SwiftUI
import SuzentCore

struct PairingView: View {
    @Bindable var model: MobileModel
    @State private var scanner = false
    @ScaledMetric(relativeTo: .title2) private var titleSize = PresentationTokens.typeTitle
    @ScaledMetric(relativeTo: .body) private var bodySize = PresentationTokens.typeBody

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: PresentationTokens.spaceLarge) {
                VStack(alignment: .leading, spacing: PresentationTokens.spaceSmall) {
                    Text("Your Suzent, together.").font(.system(size: titleSize, weight: .bold))
                    Text("Connect to your desktop. Continue conversations wherever you are.")
                }
                .padding(PresentationTokens.spaceLarge)
                .foregroundStyle(.black)
                .frame(maxWidth: .infinity, alignment: .leading)
                .background(Color(presentation: PresentationTokens.yellow))
                .overlay(Rectangle().stroke(.black, lineWidth: PresentationTokens.borderWidth))

                if let code = model.pairingCode {
                    ProgressView("Waiting for desktop approval")
                    Text(code).font(.system(.largeTitle, design: .monospaced).bold()).textSelection(.enabled)
                    Text("Check this code on your desktop, then choose which conversations and actions to share.")
                    Button("Cancel pairing", action: model.cancelPairing).buttonStyle(SuzentButtonStyle())
                } else if let invitation = model.pairingInvitation {
                    Text("Connect to this desktop?").font(.title2.bold())
                    if let preview = model.pairingPreview {
                        Text(preview.desktopName).font(.title2.bold())
                        PairingPermissionsView(permissions: preview.permissions)
                    }
                    Text(invitation.origin).font(.body.monospaced()).textSelection(.enabled)
                    Text(invitation.phoneConfirmation
                         ? String(localized: "Confirm to connect with the permissions shown above. You do not need to approve again on desktop.")
                         : String(localized: "Your desktop will ask you to approve this phone. Node access stays separate."))
                    if model.busy { ProgressView() }
                    Button(invitation.phoneConfirmation ? String(localized: "Confirm connection") : String(localized: "Request pairing"), action: model.approveDestination)
                        .buttonStyle(SuzentButtonStyle(prominent: true)).disabled(model.busy)
                    Button("Cancel", action: model.cancelPairing).disabled(model.busy)
                } else if model.busy {
                    ProgressView("Checking desktop addresses…")
                    Button("Cancel pairing", action: model.cancelPairing).buttonStyle(SuzentButtonStyle())
                } else {
                    Text("On desktop, open Settings → Devices → Mobile access, then choose Pair a phone.")
                    Button("Scan desktop QR code", systemImage: "qrcode.viewfinder") { scanner = true }
                        .buttonStyle(SuzentButtonStyle(prominent: true)).disabled(model.busy)
                    DisclosureGroup("Paste an invitation instead") {
                        VStack(alignment: .leading, spacing: PresentationTokens.spaceMedium) {
                            SuzentTextInput(placeholder: "Pairing invitation", text: $model.invitationText, multiline: true)
                            Button("Review invitation") { model.stageInvitation(model.invitationText) }
                                .buttonStyle(SuzentButtonStyle()).disabled(model.busy || model.invitationText.isEmpty)
                        }.padding(.top)
                    }
                    if model.canReconnect {
                        Divider()
                        Text(model.origin).font(.caption).foregroundStyle(.secondary)
                        Button("Reconnect to desktop") { Task { await model.connect() } }
                            .buttonStyle(SuzentButtonStyle()).disabled(model.busy)
                    }
                }
            }.padding(PresentationTokens.spacePage)
        }
        .font(.system(size: bodySize))
        .fullScreenCover(isPresented: $scanner) { PairingScanner(model: model) }
    }
}

struct ClientPermissionsView: View {
    let device: ClientDevice
    let origin: String

    var body: some View {
        VStack(alignment: .leading, spacing: PresentationTokens.spaceMedium) {
            Text(device.displayName).font(.headline)
            Text(origin).font(.caption.monospaced()).foregroundStyle(.secondary)
            Label(device.permissions.allChats ? String(localized: "All conversations")
                : String(localized: "Shared conversations: \(device.permissions.chatIds.count)"), systemImage: "text.bubble")
            permission("Create conversations", enabled: device.permissions.createChats)
            permission("Send messages", enabled: device.permissions.send)
            permission("Stop responses", enabled: device.permissions.stop)
            permission("Approve tool requests", enabled: device.permissions.approveTools == true)
            Text("Manage or revoke access in desktop Settings → Devices → Mobile access.")
                .font(.footnote).foregroundStyle(.secondary)
            Text("Tool requests follow the conversation policy. Node access is separate.")
                .font(.footnote).foregroundStyle(.secondary)
        }
    }
    private func permission(_ title: LocalizedStringKey, enabled: Bool) -> some View {
        HStack { Text(title); Spacer(); Text(enabled ? String(localized: "Allowed") : String(localized: "Not allowed")).foregroundStyle(.secondary) }
    }
}


struct PairingPermissionsView: View {
    let permissions: ClientPermissions
    var body: some View {
        VStack(alignment: .leading, spacing: PresentationTokens.spaceSmall) {
            if permissions.allChats && permissions.createChats && permissions.send && permissions.stop && permissions.approveTools == true {
                Text("Full access").font(.headline)
                Text("All conversations, messaging, and tool approvals.").font(.subheadline)
            } else {
            Text(permissions.allChats ? String(localized: "All conversations")
                 : String(localized: "Shared conversations: \(permissions.chatIds.count)"))
            permission("Create conversations", enabled: permissions.createChats)
            permission("Send messages", enabled: permissions.send)
            permission("Stop responses", enabled: permissions.stop)
            permission("Approve tool requests", enabled: permissions.approveTools == true)
            }
        }.frame(maxWidth: .infinity, alignment: .leading).padding(16)
            .overlay(Rectangle().stroke(.primary, lineWidth: PresentationTokens.borderWidth))
    }
    private func permission(_ title: LocalizedStringKey, enabled: Bool) -> some View {
        HStack { Text(title); Spacer(); Text(enabled ? String(localized: "Allowed") : String(localized: "Not allowed")) }
    }
}

File: apps/android/app/src/main/java/com/suzent/mobile/PairingViews.kt
package com.suzent.mobile

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp

@Composable
fun PairingView(model: MobileModel) {
    var scanning by remember { mutableStateOf(false) }
    val launchScanner = { scanning = true }

    if (scanning) PairingScanner(model) { scanning = false }
    var paste by remember { mutableStateOf(false) }
    Column(Modifier.fillMaxSize().padding(top = 16.dp).verticalScroll(rememberScrollState()),
        verticalArrangement = Arrangement.spacedBy(PresentationTokens.spaceLarge.dp)) {
        Column(Modifier.fillMaxWidth().background(Color(PresentationTokens.yellow))
            .border(PresentationTokens.borderWidth.dp, Color.Black).padding(PresentationTokens.spaceLarge.dp),
            verticalArrangement = Arrangement.spacedBy(PresentationTokens.spaceSmall.dp)) {
            Text(stringResource(R.string.pairing_heading), style = MaterialTheme.typography.titleLarge, color = Color.Black)
            Text(stringResource(R.string.pairing_intro), color = Color.Black)
        }
        val invitation = model.pairingInvitation
        val code = model.pairingCode
        if (code != null) {
            LinearProgressIndicator(Modifier.fillMaxWidth())
            Text(stringResource(R.string.waiting_approval), style = MaterialTheme.typography.titleLarge)
            Text(code, style = MaterialTheme.typography.headlineLarge)
            Text(stringResource(R.string.compare_code))
            SuzentAction(stringResource(R.string.cancel_pairing), model::cancelPairing)
        } else if (invitation != null) {
            Text(stringResource(R.string.confirm_desktop), style = MaterialTheme.typography.titleLarge)
            model.pairingPreview?.let { preview ->
                Text(preview.desktopName, style = MaterialTheme.typography.titleLarge)
                val permissions = preview.permissions
                Column(Modifier.fillMaxWidth().border(PresentationTokens.borderWidth.dp, MaterialTheme.colorScheme.outline).padding(16.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    if (permissions.allChats && permissions.createChats && permissions.send && permissions.stop && permissions.approveTools) {
                        Text(stringResource(R.string.full_access), style = MaterialTheme.typography.titleMedium)
                        Text(stringResource(R.string.full_access_summary))
                    } else {
                    Text(if (permissions.allChats) stringResource(R.string.all_conversations)
                        else stringResource(R.string.shared_conversations, permissions.chatIds.size))
                    listOf(R.string.create_conversations to permissions.createChats, R.string.send_messages to permissions.send,
                        R.string.stop_responses to permissions.stop, R.string.approve_tools to permissions.approveTools).forEach { (label, allowed) ->
                        Text(stringResource(label) + ": " + stringResource(if (allowed) R.string.allowed else R.string.not_allowed))
                    }
                    }
                }
            }
            Text(invitation.origin)
            Text(stringResource(if (invitation.phoneConfirmation) R.string.confirm_connection_help else R.string.confirm_desktop_help))
            if (model.busy) LinearProgressIndicator(Modifier.fillMaxWidth())
            SuzentAction(stringResource(if (invitation.phoneConfirmation) R.string.confirm_connection else R.string.request_pairing),
                model::approveDestination, prominent = true, enabled = !model.busy)
            SuzentTextButton(onClick = model::cancelPairing, enabled = !model.busy) { Text(stringResource(R.string.cancel_pairing)) }
        } else if (model.busy) {
            LinearProgressIndicator(Modifier.fillMaxWidth())
            Text(stringResource(R.string.checking_addresses))
            SuzentAction(stringResource(R.string.cancel_pairing), model::cancelPairing)
        } else {
            Text(stringResource(R.string.pairing_steps))
            SuzentAction(stringResource(R.string.scan_desktop), {
                launchScanner()
            }, prominent = true, enabled = !model.busy)
            SuzentTextButton(onClick = { paste = !paste }) { Text(stringResource(R.string.paste_invitation)) }
            if (paste) {
                SuzentTextInput(model.invitationText, { model.invitationText = it }, stringResource(R.string.pairing_invitation), multiline = true)
                SuzentAction(stringResource(R.string.review_invitation), { model.stageInvitation(model.invitationText) },
                    enabled = !model.busy && model.invitationText.isNotBlank())
            }
            if (model.canReconnect) {
                HorizontalDivider()
                Text(model.origin, style = MaterialTheme.typography.bodySmall)
                SuzentAction(stringResource(R.string.reconnect), model::connect, enabled = !model.busy)
            }
        }
        Spacer(Modifier.height(PresentationTokens.spaceLarge.dp))
    }
}

@Composable
fun ClientPermissionsView(device: ClientDevice, origin: String) {
    Column(Modifier.fillMaxWidth().border(PresentationTokens.borderWidth.dp, MaterialTheme.colorScheme.outline)
        .padding(PresentationTokens.spacePage.dp), verticalArrangement = Arrangement.spacedBy(PresentationTokens.spaceSmall.dp)) {
        Text(stringResource(R.string.desktop_access), style = MaterialTheme.typography.titleLarge)
        Text(device.name, style = MaterialTheme.typography.titleMedium)
        Text(origin, style = MaterialTheme.typography.bodySmall)
        Text(if (device.permissions.allChats) stringResource(R.string.all_conversations)
            else stringResource(R.string.shared_conversations, device.permissions.chatIds.size))
        listOf(R.string.create_conversations to device.permissions.createChats,
            R.string.send_messages to device.permissions.send, R.string.stop_responses to device.permissions.stop, R.string.approve_tools to device.permissions.approveTools).forEach { (label, enabled) ->
            Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                Text(stringResource(label))
                Text(stringResource(if (enabled) R.string.allowed else R.string.not_allowed), color = MaterialTheme.colorScheme.onSurfaceVariant)
            }
        }
        Text(stringResource(R.string.manage_access), style = MaterialTheme.typography.bodySmall)
        Text(stringResource(R.string.tool_approval_desktop), style = MaterialTheme.typography.bodySmall)
    }
}
